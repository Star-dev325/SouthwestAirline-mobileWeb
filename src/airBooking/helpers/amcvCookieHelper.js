// @flow
import _ from 'lodash';
import * as Cookie from 'src/shared/swa-persistence/cookie';

import { buildPathWithQuery } from 'src/shared/helpers/pathUtils';

const AMCV_COOKIE_NAME = 'AMCV_65D316D751E563EC0A490D4C%40AdobeOrg';
const AMCV_COOKIE_KEY_SEPARATOR = 'MID';
const AMCV_COOKIE_PIPE_SEPARATOR = '|';
const CHASE_URL_KEYWORD = 'creditcard';

export type ChaseAppendParams = {
  isChaseCombo?: boolean,
  pageId?: string
};

export const getMcvid = () => {
  const value = Cookie.getValue(AMCV_COOKIE_NAME);

  return _.chain(value).split(AMCV_COOKIE_KEY_SEPARATOR).nth(1).split(AMCV_COOKIE_PIPE_SEPARATOR).nth(1).value() || '';
};

// Deprecated - please use appendParamsIfChaseUrl instead
export const addMvcidToChaseUrl = (url: string, isChase: boolean) => {
  let updateUrl = url;

  if (!_.isEmpty(url) && (isChase || url.includes(CHASE_URL_KEYWORD))) {
    const mcvid = getMcvid();

    if (!_.isEmpty(mcvid)) {
      updateUrl = `${url}&mcvid=${mcvid}`;
    }
  }

  return updateUrl;
};

export const appendParamsIfChaseUrl = (target: string = '', params: ChaseAppendParams = {}) => {
  const { isChaseCombo = false, pageId = '' } = params;
  const query = {};
  const isChaseUrl = _.includes(target, CHASE_URL_KEYWORD) || isChaseCombo;
  const mcvid = isChaseUrl && getMcvid();

  mcvid && _.set(query, 'mcvid', mcvid);
  pageId && _.set(query, 'pageId', pageId);

  return buildPathWithQuery(target, query);
};
