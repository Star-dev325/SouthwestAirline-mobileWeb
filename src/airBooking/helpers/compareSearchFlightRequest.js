import _ from 'lodash';
import { transformToAPIRequest } from 'src/shared/transformers/flightProductSearchRequestTransformer';

export const compareSearchFlightRequest = (requestOne, requestTwo) => {
  const transformedRequestOneQuery = _.get(transformToAPIRequest(requestOne), 'query');
  const transformedRequestTwoQuery = _.get(transformToAPIRequest(requestTwo), 'query');
  const requestOneWithoutPromoCode = _.omit(transformedRequestOneQuery, 'promo-code');
  const requestTwoWithoutPromoCode = _.omit(transformedRequestTwoQuery, 'promo-code');

  return _.isEqual(requestOneWithoutPromoCode, requestTwoWithoutPromoCode);
};
