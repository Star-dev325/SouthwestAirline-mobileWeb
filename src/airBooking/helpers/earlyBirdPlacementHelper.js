// @flow

import _ from 'lodash';
import numeral from 'numeral';
import i18n from '@swa-ui/locale';
import type { EarlyBirdEligibility } from 'src/airBooking/flow-typed/airBooking.types';
import type { CurrencyType } from 'src/shared/flow-typed/shared.types';

const getEarlyBirdPrefixText = (unitPriceInBound: CurrencyType, unitPriceOutBound: CurrencyType) => {
  const showStartingFromMessage =
    !!unitPriceOutBound &&
    !!unitPriceInBound &&
    _.get(unitPriceOutBound, 'amount') !== _.get(unitPriceInBound, 'amount');

  return showStartingFromMessage ? i18n('SHARED__EARLY_BIRD__PER_PASSENGER_PRICE_MESSAGE_PREFIX') : '(';
};

const getEarlyBirdUnitAmount = (unitPriceInBound: CurrencyType, unitPriceOutBound: CurrencyType) => {
  const earlyBirdUnitPrice = _.chain([unitPriceOutBound, unitPriceInBound])
    .compact()
    .minBy((item) => numeral(item.amount).value())
    .value();

  return _.get(earlyBirdUnitPrice, 'amount');
};

export const getEarlyBirdAdditionalTemplateData = (
  unitPriceInBound: CurrencyType,
  unitPriceOutBound: CurrencyType,
  earlyBirdEligibility: ?EarlyBirdEligibility
) => ({
  earlyBirdTotalPrice: _.get(earlyBirdEligibility, 'totalPrice.amount'),
  earlyBirdUnitPrice: getEarlyBirdUnitAmount(unitPriceInBound, unitPriceOutBound),
  earlyBirdPrefixText: getEarlyBirdPrefixText(unitPriceInBound, unitPriceOutBound)
});
