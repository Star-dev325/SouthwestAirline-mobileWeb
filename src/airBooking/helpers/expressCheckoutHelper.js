// @flow
import _ from 'lodash';
import type { Passenger } from 'src/shared/flow-typed/shared.types';
import validator from 'src/shared/form/formValidators/validator';

export const isInformationCompletedForExpressCheckout = (passengerDetailsPage: Passenger): boolean => {
  const { firstName, lastName, dateOfBirth, gender } = passengerDetailsPage || {};

  return !_.isEmpty(firstName) && !_.isEmpty(lastName) && !_.isEmpty(dateOfBirth) && validator.isValidGender(gender);
};
