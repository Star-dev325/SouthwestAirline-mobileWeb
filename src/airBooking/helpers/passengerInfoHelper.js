// @flow

import dayjs from 'dayjs';

export const formatDateOfBirthToYearMonthDay = (dateOfBirth: string) =>
  dayjs(dateOfBirth, ['MM/DD/YYYY', 'YYYY-MM-DD']).format('YYYY-MM-DD');

export const formatDateOfBirthToMonthDayYear = (dateOfBirth: string) =>
  dayjs(dateOfBirth, ['MM/DD/YYYY', 'YYYY-MM-DD']).format('MM/DD/YYYY');
