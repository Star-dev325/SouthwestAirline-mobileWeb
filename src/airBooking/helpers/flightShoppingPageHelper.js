import _ from 'lodash';
import { permutations } from 'src/shared/helpers/mathUtils';
import { INBOUND, OUTBOUND } from 'src/shared/constants/flightBoundTypes';
import PassengerTypes from 'src/shared/constants/passengerTypes';

const { ADULT } = PassengerTypes;

export const getNextFlightShoppingPageParams = (params, response) => {
  const { inboundPage } = response;
  const bounds = inboundPage ? [OUTBOUND, INBOUND] : [OUTBOUND];
  const paxTypes = [ADULT];

  const pages = permutations(paxTypes, bounds).map((paxAndBound) => ({
    paxType: paxAndBound[0],
    direction: paxAndBound[1]
  }));

  const index = _.findIndex(pages, params);

  return index === -1 ? undefined : _.get(pages, `[${index + 1}]`);
};

export const getFirstShoppingPageParams = () => ({
  paxType: ADULT,
  direction: OUTBOUND
});

export const generateFlightShoppingPages = (response) => {
  const { outboundPage, inboundPage } = response.flightShoppingPage;
  const isRoundTrip = !!inboundPage;
  let pages = [];

  const buildPages = () => {
    const result = [
      {
        direction: OUTBOUND,
        paxType: ADULT,
        ...outboundPage
      }
    ];

    if (isRoundTrip) {
      result.push({
        direction: INBOUND,
        paxType: ADULT,
        ...inboundPage
      });
    }

    return result;
  };

  pages = pages.concat(buildPages());

  return pages;
};
