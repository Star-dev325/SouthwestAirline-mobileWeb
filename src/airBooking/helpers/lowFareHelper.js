// @flow
import _ from 'lodash';
import dayjs from 'dayjs';
import type {
  LowFareCalendarDaysType,
  LowFareByMonthArraysType,
  LowFareCalendarPageType
} from 'src/airBooking/flow-typed/lowFare.types';
import { convertToNumber } from 'src/shared/helpers/numberHelper';
import { isPastDate } from 'src/shared/helpers/dateHelper';

export const MIN_BAR_HEIGHT_WITH_TAXES = 4.5;
export const MIN_BAR_HEIGHT = 3.7;
export const MAX_BAR_HEIGHT = 19;

const SMOOTH_SCROLL_TIME = 300;
const REM_CONVERSION = 10;

export const hasPricePoints = (boundFlights: Array<LowFareCalendarDaysType>) =>
  !_.chain(boundFlights)
    .map((fare) => !_.isEmpty(_.get(fare, 'lowestPrice.pricePointsTax')))
    .uniq()
    .compact()
    .isEmpty()
    .value();

export const getMinPrice = (boundFlights: Array<LowFareCalendarDaysType>, todayAsDate: string) =>
  _.chain(boundFlights)
    .map((fare) => {
      const value = _.get(fare, 'lowestPrice.price.amount', undefined);
      const amount = convertToNumber(value);
      const min = isPastDate(fare.date, todayAsDate) ? undefined : amount;

      return min;
    })
    .min()
    .value() || 0;

export const getMaxPrice = (boundFlights: Array<LowFareCalendarDaysType>, todayAsDate: string) =>
  _.chain(boundFlights)
    .map((fare) => {
      const value = _.get(fare, 'lowestPrice.price.amount', 0);
      const amount = convertToNumber(value);

      return isPastDate(fare.date, todayAsDate) ? 0 : amount;
    })
    .max()
    .value() || 0;

const percentageInPriceRange = (farePrice: number, minPrice: number, maxPrice: number): number => {
  const denom: number = maxPrice - minPrice;

  if (denom === 0) {
    return 1;
  }

  return (farePrice - minPrice) / denom;
};

export const getBarHeight = (
  farePrice: ?number,
  minPrice: number,
  maxPrice: number,
  showPointsTax: boolean
): string => {
  const minBarHeight = showPointsTax ? MIN_BAR_HEIGHT_WITH_TAXES : MIN_BAR_HEIGHT;

  if (!maxPrice || !minPrice || farePrice === 0 || farePrice === undefined || farePrice === null) {
    return `${minBarHeight}rem`;
  }

  const percentage = percentageInPriceRange(farePrice, minPrice, maxPrice);
  const adjustedHeight = MAX_BAR_HEIGHT * percentage + minBarHeight * (1 - percentage);

  return `${adjustedHeight}rem`;
};

export const getLowFareDaysByMonth = (boundFares: Array<LowFareCalendarDaysType>): LowFareByMonthArraysType => {
  const lowFareDaysMonthMap = new Map();
  let faresInThisMonth: Array<LowFareCalendarDaysType> = [];
  let currentMonthYear: string; // 1 = Jan ... 12 = Dec  Ex. 1-2019

  _.forEach(boundFares, (fare: LowFareCalendarDaysType) => {
    const thisDate = dayjs(fare.date);

    currentMonthYear = thisDate.format('M-YYYY');

    faresInThisMonth = lowFareDaysMonthMap.get(currentMonthYear) || [];
    faresInThisMonth.push(fare);
    lowFareDaysMonthMap.set(currentMonthYear, faresInThisMonth);
  });

  const lowFareDaysMonthArray: LowFareByMonthArraysType = [];

  for (const value of lowFareDaysMonthMap.values()) {
    lowFareDaysMonthArray.push(value);
  }

  return lowFareDaysMonthArray;
};

export const calculateLfcSelectedDates = (searchRequest: *, response: LowFareCalendarPageType) => {
  const { departureDate, returnDate } = searchRequest;
  const outboundDays = _.get(response, 'lowFareCalendarPage.outboundPage.lowFareCalendarDays');
  const inboundDays = _.get(response, 'lowFareCalendarPage.inboundPage.lowFareCalendarDays');
  let selectedDepartureDate;
  let selectedReturnDate;

  if (
    departureDate &&
    outboundDays &&
    outboundDays.filter((d) => d.date === departureDate && d.lowestPrice).length > 0
  ) {
    selectedDepartureDate = departureDate;
  }

  if (returnDate && inboundDays && inboundDays.filter((d) => d.date === returnDate && d.lowestPrice).length > 0) {
    selectedReturnDate = returnDate;
  }

  return { selectedDepartureDate, selectedReturnDate };
};

export const calculateCalendarScrollableBounds = (calendar: *, leftSpacer: *, rightSpacer: *) => {
  const calDays = calendar.getElementsByClassName('calendar-day');
  const firstBar = calDays[0];
  const lastBar = calDays[calDays.length - 1];
  const outboundCalRefRect = calendar.getBoundingClientRect();
  const firstBarRect = firstBar.getBoundingClientRect();
  const lastBarRect = lastBar.getBoundingClientRect();
  const calMidPoint = outboundCalRefRect.left + outboundCalRefRect.width / 2;
  const diffLeftToCenter = Math.round(calMidPoint - (firstBar.offsetLeft + firstBarRect.width / 2));
  const diffRightToCenter = Math.round(
    calMidPoint - (calendar.scrollWidth - (lastBar.offsetLeft + lastBarRect.width / 2))
  );
  const calendarPaddingLeft = diffLeftToCenter / REM_CONVERSION;
  const calendarPaddingRight = diffRightToCenter / REM_CONVERSION;

  if (calendarPaddingLeft) {
    const currentPaddingLeft = leftSpacer.style['paddingLeft']
      ? parseFloat(leftSpacer.style['paddingLeft'].replace('rem', ''))
      : 0;

    leftSpacer.style['paddingLeft'] = `${currentPaddingLeft + calendarPaddingLeft}rem`;
  }

  if (calendarPaddingRight) {
    const currentPaddingRight = rightSpacer.style['paddingRight']
      ? parseFloat(rightSpacer.style['paddingRight'].replace('rem', ''))
      : 0;

    rightSpacer.style['paddingRight'] = `${currentPaddingRight + calendarPaddingRight}rem`;
  }
};

export const calendarScrollFunctions = {
  focusOnPriceBar: (lowFareCalendar: *, targetEl: HTMLElement, noAnimation?: boolean) => {
    if (lowFareCalendar) {
      const lfcClientRect = lowFareCalendar.getBoundingClientRect();
      const targetClientRect = targetEl.getBoundingClientRect();
      const differenceFromCenter = Math.round(
        targetClientRect.left + targetClientRect.width / 2 - (lfcClientRect.left + lfcClientRect.width / 2)
      );
      const positionToScrollTo = lowFareCalendar.scrollLeft + differenceFromCenter;
      const startTime = _.has(window, 'performance.now') ? performance.now() : new Date().getTime();

      calendarScrollFunctions.scrollToPriceBar(
        noAnimation,
        lowFareCalendar,
        positionToScrollTo,
        startTime,
        lowFareCalendar.scrollLeft
      );
    }
  },
  scrollToPriceBar: (
    skipAnimation: boolean = false,
    elementToScroll: HTMLElement = document.createElement('span'),
    positionToScrollTo: number = 0,
    startTime: number = 0,
    startPosition: number = 0
  ) => {
    if (skipAnimation) {
      elementToScroll.scrollLeft = positionToScrollTo;

      return;
    }

    const currentTime = _.has(window, 'performance.now') ? performance.now() : new Date().getTime();
    const progress = (currentTime - startTime) / SMOOTH_SCROLL_TIME;
    const percentageComplete = Math.min(1, progress);
    const newPosition = Math.ceil(
      easeOutQuad(percentageComplete) * (positionToScrollTo - startPosition) + startPosition
    );

    if (elementToScroll.scrollLeft === positionToScrollTo || isAtEndOfScroll(elementToScroll, percentageComplete)) {
      return; // end scrolling
    }

    elementToScroll.scrollLeft = newPosition;

    if (_.has(window, 'requestAnimationFrame')) {
      window.requestAnimationFrame(
        calendarScrollFunctions.scrollToPriceBar.bind(
          null,
          false,
          elementToScroll,
          positionToScrollTo,
          startTime,
          startPosition
        )
      );
    }
  }
};

const easeOutQuad = (t: number) => t * (2 - t);

const isAtEndOfScroll = (elementToScroll: HTMLElement, percentageComplete: number) => {
  if (percentageComplete !== 1) {
    return false;
  }

  return elementToScroll.scrollLeft === 0 ||
    elementToScroll.offsetWidth + elementToScroll.scrollLeft === elementToScroll.scrollWidth ||
    percentageComplete >= 1;
};
