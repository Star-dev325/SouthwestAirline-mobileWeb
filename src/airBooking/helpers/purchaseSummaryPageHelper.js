// @flow
import i18n from '@swa-ui/locale';
import dayjs from 'dayjs';
import _ from 'lodash';
import type { FlightPricingPageResponse, TripSummaryType } from 'src/airBooking/flow-typed/airBooking.types';
import { TRAVEL_MANAGER_OPTIONS } from 'src/shared/constants/contactMethodOptions';
import { MEDIUM_DATE_FORMAT } from 'src/shared/constants/dateConstants';
import PassengerTypes from 'src/shared/constants/passengerTypes';
import optionsHelper from 'src/shared/helpers/optionsHelper';

const { LAPCHILD } = PassengerTypes;

import type {
  CurrencySuit,
  DutyOfCare,
  EarlyBirdPricing,
  Passenger,
  PassengerInfos,
  TotalsType
} from 'src/shared/flow-typed/shared.types';
import { _addHyphenForUSPhoneNumber } from 'src/shared/helpers/contactMethodPageHelper';

type BriefPassengerInfoType = {
  name: string,
  rapidRewardsNumber?: string
};

const contactMethodKeys = optionsHelper.keyMirror(TRAVEL_MANAGER_OPTIONS);

export const generateTripSummary = (flightPricingPageResponse: FlightPricingPageResponse): TripSummaryType => {
  const lapChildEnabled = _.get(flightPricingPageResponse, 'flightPricingPage.totals.infantFare');
  const bounds = _.chain(flightPricingPageResponse)
    .get('flightPricingPage.bounds')
    .map((bound) => ({
      arrivalAirportCode: bound.arrivalAirport.code,
      arrivalTime: bound.arrivalTime,
      departureAirportCode: bound.departureAirport.code,
      departureDate: bound.departureDate,
      departureDayOfWeek: dayjs(bound.departureDate).format('dddd'),
      departureTime: bound.departureTime,
      stops: bound.stops
    }))
    .value();

  const outboundData = _.get(flightPricingPageResponse, 'flightPricingPage.bounds[0].passengers', [
    {
      count: 0,
      type: i18n('SHARED__PURCHASE_SUMMARY_FORM__DEFAULT_PASSENGER_TYPE')
    }
  ]);
  const defaultCurrency = {
    amount: '0.00',
    currencyCode: 'USD',
    currencySymbol: '$'
  };

  const pointsTotal = _.get(flightPricingPageResponse, 'flightPricingPage.totals.pointsTotal');
  const adultFareMoneyTotal = _.get(
    flightPricingPageResponse,
    'flightPricingPage.totals.adultFare.paxTypeTotal.moneyTotal',
    defaultCurrency
  );
  const infantFareMoneyTotal = _.get(
    flightPricingPageResponse,
    'flightPricingPage.totals.infantFare.paxTypeTotal.moneyTotal',
    defaultCurrency
  );
  const currency = pointsTotal || adultFareMoneyTotal;
  const defaultLapChildCurrency = pointsTotal || infantFareMoneyTotal;

  return {
    bounds,
    currency,
    ...(lapChildEnabled && { defaultLapChildCurrency }),
    passengerCountDescription: `${outboundData[0].count} ${outboundData[0].type} ${i18n(
      'SHARED__PURCHASE_SUMMARY_FORM__TOTAL'
    )}`,
    ...(lapChildEnabled && {
      lapChildCountDescription: `${outboundData[1].count} ${outboundData[1].type} ${i18n(
        'SHARED__PURCHASE_SUMMARY_FORM__TOTAL'
      )}`
    })
  };
};

export const generatePassengers = (
  passengerInfos: Array<{ passengerInfo: Passenger }>
): Array<BriefPassengerInfoType> =>
  _.map(passengerInfos, (passenger) => {
    const briefInfo: BriefPassengerInfoType = {
      name: `${passenger.passengerInfo.firstName} ${passenger.passengerInfo.lastName}`
    };

    if (passenger.passengerInfo.rapidRewardsNumber) {
      briefInfo.rapidRewardsNumber = passenger.passengerInfo.rapidRewardsNumber;
    }

    return briefInfo;
  });

export const generatePriceTotal = (flightPricingPageResponse: FlightPricingPageResponse): { totals: TotalsType } => {
  const totals = _.get(flightPricingPageResponse, 'flightPricingPage.totals');

  return {
    totals
  };
};

export const getReviewMessages = (flightPricingPageResponse: FlightPricingPageResponse) =>
  flightPricingPageResponse?.flightPricingPage?.reviewMessages || null;

const transformPassengersForEarlybirdRequest = (productIds: string[], passengerInfos: PassengerInfos) => ({
  productIds,
  passengers: _.map(passengerInfos, ({ passengerInfo, passengerReference }) => {
    const result = {
      name: _.pick(passengerInfo, ['firstName', 'lastName', 'middleName']),
      passengerReference,
      gender: passengerInfo.gender,
      dateOfBirth: passengerInfo.dateOfBirth
    };

    if (!_.isEmpty(passengerInfo.rapidRewardsNumber)) {
      _.merge(result, { accountNumber: passengerInfo.rapidRewardsNumber });
    }

    if (!_.isEmpty(passengerInfo.frequentTravelerId)) {
      _.merge(result, { frequentTravelerId: passengerInfo.frequentTravelerId });
    }

    if (!_.isEmpty(passengerInfo.frequentTravelerToken)) {
      _.merge(result, { frequentTravelerToken: passengerInfo.frequentTravelerToken });
    }

    return result;
  })
});

export const transformToEarlybirdInPathRequest = (
  earlyBirdPricing: EarlyBirdPricing,
  passengerInfos: PassengerInfos,
  currencyType: CurrencySuit,
  earlyBirdPricingToken?: string
) => {
  const earlybirdInPathRequest = {
    body: {
      currency: currencyType,
      ..._.omitIfEmpty({ earlyBirdPricingToken })
    },
    method: earlyBirdPricing.method,
    href: earlyBirdPricing.href
  };

  _.each({ adultPassengers: 'adult' }, (paxType, paxKey) => {
    const passengers = _.filter(passengerInfos, { type: paxType });
    const productIds = _.get(earlyBirdPricing, `body.${paxKey}.productIds`);

    if (!_.isEmpty(passengers) && !_.isEmpty(productIds)) {
      _.merge(earlybirdInPathRequest.body, {
        [paxKey]: transformPassengersForEarlybirdRequest(productIds, passengers)
      });
    }
  });

  return earlybirdInPathRequest;
};

export const getContactInfoTravelManagerText = ({ contactMethod, contactPhone, contactEmail }: DutyOfCare) => {
  if (contactMethod === contactMethodKeys.CALL_ME && contactPhone) {
    return `(${contactPhone.countryCode}) ${_addHyphenForUSPhoneNumber(
      contactPhone?.countryCode,
      contactPhone?.number
    )}`;
  } else if (contactMethod === contactMethodKeys.EMAIL_ME) {
    return contactEmail;
  }

  return '';
};

export const otherPassengerReference = (passengerType: string, passengerReference: number, associatedAdult?: string) =>
  (passengerType === LAPCHILD && associatedAdult === '' ? passengerReference - 1 : parseInt(associatedAdult));

const adultStartAge = 18;
const youngTravelerMaxAge = 17;
const youngTravelerStartAge = 12;

export const isValidYoungTravelerBirthDate = (birthDate: string, departureDate: string) =>
  dayjs(departureDate, MEDIUM_DATE_FORMAT).diff(dayjs(birthDate, [MEDIUM_DATE_FORMAT]), 'years') >= youngTravelerStartAge &&
  dayjs(departureDate, MEDIUM_DATE_FORMAT).diff(dayjs(birthDate, [MEDIUM_DATE_FORMAT]), 'years') <= youngTravelerMaxAge;

export const isValidAdultBirthDate = (birthDate: string, departureDate: string) =>
  dayjs(departureDate, MEDIUM_DATE_FORMAT).diff(dayjs(birthDate, [MEDIUM_DATE_FORMAT]), 'years') >= adultStartAge;
