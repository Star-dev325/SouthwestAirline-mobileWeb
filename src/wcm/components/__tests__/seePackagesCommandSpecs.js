import React from 'react';
import { shallow } from 'enzyme';
import BoundDetailBuilder from 'test/builders/model/boundDetailBuilder';
import SearchForFlightsRequestBuilder from 'test/builders/model/searchForFlightsRequestBuilder';
import proxyquire from 'proxyquire';
import { sandbox } from 'sinon';
import browser from 'src/shared/helpers/browserObject';
import { click } from 'test/unit/helpers/enzymeFormTestUtils';

const sinon = sandbox.create();

describe('See packages command', () => {
  const mockSwaVacationsUrl = 'https://res.southwestvacations.com/search/ExternalFormPost.aspx?cmpid=SWA-INPATH';
  const mockProps = {
    bounds: [new BoundDetailBuilder().build(), new BoundDetailBuilder().build()],
    searchRequest: new SearchForFlightsRequestBuilder().build(),
    commandProps: {
      utm_campaign: 'mockValue',
      utm_medium: 'mockValue',
      utm_source: 'mockValue',
      utm_term: 'mockValue',
      utm_content: 'mockValue'
    }
  };
  let mockPost, openInBrowserStub;
  let SeePackagesCommand;

  beforeEach(() => {
    mockPost = sinon.stub();
    openInBrowserStub = sinon.spy(browser.window, 'open');
    SeePackagesCommand = proxyquire('src/wcm/components/seePackagesCommand', {
      'src/shared/form/helpers/post': { default: mockPost },
      'src/shared/api/apiRoutes': {
        default: {
          swaVacationsUrl: mockSwaVacationsUrl
        }
      }
    }).default;
  });

  afterEach(() => {
    sandbox.restore();
    browser.window.open.restore();
  });

  it('should should open default url', () => {
    const wrapper = createComponent();

    click(wrapper.find('.see-packages'));

    expect(openInBrowserStub.args[0]).toMatchSnapshot();
  });

  it('should should open default url when isWebView is true', () => {
    const wrapper = createComponent({ isWebView: true });

    click(wrapper.find('.see-packages'));

    expect(openInBrowserStub.args[0]).toMatchSnapshot();
  });

  it('should send additional parameters in path url', () => {
    const updatedCommandProps = { ...mockProps.commandProps, additionalParam: 'mockValue' };
    const wrapper = createComponent({ commandProps: updatedCommandProps });

    click(wrapper.find('.see-packages'));

    expect(openInBrowserStub.args[0]).toMatchSnapshot();
  });

  it('should render the proper values for international flight', () => {
    const updatedBounds = [
      new BoundDetailBuilder().withInternationalDepartureFlight().build(),
      new BoundDetailBuilder().build()
    ];
    const wrapper = createComponent({ bounds: updatedBounds });

    click(wrapper.find('.see-packages'));

    expect(openInBrowserStub.args[0]).toMatchSnapshot();
  });

  it('should render the proper values for international flight when isWebView true', () => {
    const updatedBounds = [
      new BoundDetailBuilder().withInternationalDepartureFlight().build(),
      new BoundDetailBuilder().build()
    ];
    const wrapper = createComponent({ bounds: updatedBounds, isWebView: true });

    click(wrapper.find('.see-packages'));

    expect(openInBrowserStub.args[0]).toMatchSnapshot();
  });

  function createComponent(props = {}) {
    return shallow(
      <SeePackagesCommand {...mockProps} {...props}>
        {'test content'}
      </SeePackagesCommand>
    );
  }
});
