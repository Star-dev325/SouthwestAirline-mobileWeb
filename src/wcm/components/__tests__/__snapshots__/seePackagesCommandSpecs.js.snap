// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`See packages command should render the proper values for international flight 1`] = `
Array [
  "https://res.southwestvacations.com/search/ExternalFormPost.aspx?cmpid=SWA-INPATH&utm_campaign=mockValue&utm_medium=mockValue&utm_source=mockValue&utm_term=mockValue&utm_content=D!intpos&cpmid=SWA-INPATH&currentculture=&foPrefFlightInfo=WN1504~WGA%7CWN1504~WGA&gsAge1=&gsAge2=&gsAge3=&gsAge4=&gsDepartureDate=2015-11-10&gsDestination=AUS&gsLengthOfStay=30&gsNumberOfTravelers=1&gsOrigin=ATL&gsReturnDate=2015-12-10&gssourcecode=INPATH&gsVacationType=AH01&gsVendor=WNI&plcode=SWAVINTL",
]
`;

exports[`See packages command should render the proper values for international flight when isWebView true 1`] = `
Array [
  "https://res.southwestvacations.com/search/ExternalFormPost.aspx?cmpid=SWA-INPATH&utm_campaign=mockValue&utm_medium=mockValue&utm_source=mockValue&utm_term=app&utm_content=D!appINT&cpmid=SWA-INPATH&currentculture=&foPrefFlightInfo=WN1504~WGA%7CWN1504~WGA&gsAge1=&gsAge2=&gsAge3=&gsAge4=&gsDepartureDate=2015-11-10&gsDestination=AUS&gsLengthOfStay=30&gsNumberOfTravelers=1&gsOrigin=ATL&gsReturnDate=2015-12-10&gssourcecode=INPATH&gsVacationType=AH01&gsVendor=WNI&plcode=SWAVINTL",
]
`;

exports[`See packages command should send additional parameters in path url 1`] = `
Array [
  "https://res.southwestvacations.com/search/ExternalFormPost.aspx?cmpid=SWA-INPATH&utm_campaign=mockValue&utm_medium=mockValue&utm_source=mockValue&utm_term=mockValue&utm_content=D!mobile&additionalParam=mockValue&cpmid=SWA-INPATH&currentculture=&foPrefFlightInfo=WN1504~WGA%7CWN1504~WGA&gsAge1=&gsAge2=&gsAge3=&gsAge4=&gsDepartureDate=2015-11-10&gsDestination=AUS&gsLengthOfStay=30&gsNumberOfTravelers=1&gsOrigin=ATL&gsReturnDate=2015-12-10&gssourcecode=INPATH&gsVacationType=AH08&gsVendor=WNT&plcode=",
]
`;

exports[`See packages command should should open default url 1`] = `
Array [
  "https://res.southwestvacations.com/search/ExternalFormPost.aspx?cmpid=SWA-INPATH&utm_campaign=mockValue&utm_medium=mockValue&utm_source=mockValue&utm_term=mockValue&utm_content=D!mobile&cpmid=SWA-INPATH&currentculture=&foPrefFlightInfo=WN1504~WGA%7CWN1504~WGA&gsAge1=&gsAge2=&gsAge3=&gsAge4=&gsDepartureDate=2015-11-10&gsDestination=AUS&gsLengthOfStay=30&gsNumberOfTravelers=1&gsOrigin=ATL&gsReturnDate=2015-12-10&gssourcecode=INPATH&gsVacationType=AH08&gsVendor=WNT&plcode=",
]
`;

exports[`See packages command should should open default url when isWebView is true 1`] = `
Array [
  "https://res.southwestvacations.com/search/ExternalFormPost.aspx?cmpid=SWA-INPATH&utm_campaign=mockValue&utm_medium=mockValue&utm_source=mockValue&utm_term=app&utm_content=D!app&cpmid=SWA-INPATH&currentculture=&foPrefFlightInfo=WN1504~WGA%7CWN1504~WGA&gsAge1=&gsAge2=&gsAge3=&gsAge4=&gsDepartureDate=2015-11-10&gsDestination=AUS&gsLengthOfStay=30&gsNumberOfTravelers=1&gsOrigin=ATL&gsReturnDate=2015-12-10&gssourcecode=INPATH&gsVacationType=AH08&gsVendor=WNT&plcode=",
]
`;
